File: qt/gen_qstringconverter_base.cpp
// +build ignore

#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStringConverter>
#include <QStringConverterBase>
#define WORKAROUND_INNER_CLASS_DEFINITION_QStringConverterBase__State
#include <qstringconverter_base.h>
#include "gen_qstringconverter_base.h"

#ifndef _Bool
#define _Bool bool
#endif

QStringConverterBase* QStringConverterBase_new(QStringConverterBase* param1) {
	return new QStringConverterBase(*param1);
}

QStringConverterBase* QStringConverterBase_new2() {
	return new QStringConverterBase();
}

void QStringConverter_virtbase(QStringConverter* src, QStringConverterBase** outptr_QStringConverterBase) {
	*outptr_QStringConverterBase = static_cast<QStringConverterBase*>(src);
}

bool QStringConverter_IsValid(const QStringConverter* self) {
	return self->isValid();
}

void QStringConverter_ResetState(QStringConverter* self) {
	self->resetState();
}

bool QStringConverter_HasError(const QStringConverter* self) {
	return self->hasError();
}

const char* QStringConverter_Name(const QStringConverter* self) {
	return (const char*) self->name();
}

const char* QStringConverter_NameForEncoding(Encoding e) {
	return (const char*) QStringConverter::nameForEncoding(e);
}

struct miqt_array /* of struct miqt_string */  QStringConverter_AvailableCodecs() {
	QStringList _ret = QStringConverter::availableCodecs();
	// Convert QList<> from C++ memory to manually-managed C memory
	struct miqt_string* _arr = static_cast<struct miqt_string*>(malloc(sizeof(struct miqt_string) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		QString _lv_ret = _ret[i];
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray _lv_b = _lv_ret.toUtf8();
		struct miqt_string _lv_ms;
		_lv_ms.len = _lv_b.length();
		_lv_ms.data = static_cast<char*>(malloc(_lv_ms.len));
		memcpy(_lv_ms.data, _lv_b.data(), _lv_ms.len);
		_arr[i] = _lv_ms;
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

QStringConverterBase__State* QStringConverterBase__State_new() {
	return new QStringConverterBase::State();
}

QStringConverterBase__State* QStringConverterBase__State_new2(Flags f) {
	return new QStringConverterBase::State(f);
}

void QStringConverterBase__State_Clear(QStringConverterBase__State* self) {
	self->clear();
}

void QStringConverterBase__State_Reset(QStringConverterBase__State* self) {
	self->reset();
}

void QStringConverterBase__State_Delete(QStringConverterBase__State* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QStringConverterBase::State*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qstringconverter_base.go
package qt

import (
	"unsafe"
)

type QStringConverterBase__Flag int

const (
	QStringConverterBase__Default              QStringConverterBase__Flag = 0
	QStringConverterBase__Stateless            QStringConverterBase__Flag = 1
	QStringConverterBase__ConvertInvalidToNull QStringConverterBase__Flag = 2
	QStringConverterBase__WriteBom             QStringConverterBase__Flag = 4
	QStringConverterBase__ConvertInitialBom    QStringConverterBase__Flag = 8
	QStringConverterBase__UsesIcu              QStringConverterBase__Flag = 16
)

type QStringConverter__Encoding int

const (
	QStringConverter__Utf8         QStringConverter__Encoding = 0
	QStringConverter__Utf16        QStringConverter__Encoding = 1
	QStringConverter__Utf16LE      QStringConverter__Encoding = 2
	QStringConverter__Utf16BE      QStringConverter__Encoding = 3
	QStringConverter__Utf32        QStringConverter__Encoding = 4
	QStringConverter__Utf32LE      QStringConverter__Encoding = 5
	QStringConverter__Utf32BE      QStringConverter__Encoding = 6
	QStringConverter__Latin1       QStringConverter__Encoding = 7
	QStringConverter__System       QStringConverter__Encoding = 8
	QStringConverter__LastEncoding QStringConverter__Encoding = 8
)

type QStringConverterBase struct {
	h          uintptr
	isSubclass bool
}

// NewQStringConverterBase constructs a new QStringConverterBase object.
func NewQStringConverterBase(param1 *QStringConverterBase) *QStringConverterBase {

	ret := newQStringConverterBase(QStringConverterBase_new(param1.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQStringConverterBase2 constructs a new QStringConverterBase object.
func NewQStringConverterBase2() *QStringConverterBase {

	ret := newQStringConverterBase(QStringConverterBase_new2())
	ret.isSubclass = true
	return ret
}

type QStringConverter struct {
	h          uintptr
	isSubclass bool
}

func (this *QStringConverter) IsValid() bool {
	return (bool)(QStringConverter_IsValid(this.h))
}

func (this *QStringConverter) ResetState() {
	QStringConverter_ResetState(this.h)
}

func (this *QStringConverter) HasError() bool {
	return (bool)(QStringConverter_HasError(this.h))
}

func (this *QStringConverter) Name() string {
	_ret := QStringConverter_Name(this.h)
	return GoString(_ret)
}

func QStringConverter_NameForEncoding(e Encoding) string {
	_ret := QStringConverter_NameForEncoding(e)
	return GoString(_ret)
}

func QStringConverter_AvailableCodecs() []string {
	var _ma struct_miqt_array = QStringConverter_AvailableCodecs()
	_ret := make([]string, int(_ma.len))
	_outCast := (*[0xffff]struct_miqt_string)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		var _lv_ms struct_miqt_string = _outCast[i]
		_lv_ret := GoStringN(_lv_ms.data, int(int64(_lv_ms.len)))
		free(unsafe.Pointer(_lv_ms.data))
		_ret[i] = _lv_ret
	}
	return _ret
}

type QStringConverterBase__State struct {
	h          uintptr
	isSubclass bool
}

// NewQStringConverterBase__State constructs a new QStringConverterBase::State object.
func NewQStringConverterBase__State() *QStringConverterBase__State {

	ret := newQStringConverterBase__State(QStringConverterBase__State_new())
	ret.isSubclass = true
	return ret
}

// NewQStringConverterBase__State2 constructs a new QStringConverterBase::State object.
func NewQStringConverterBase__State2(f Flags) *QStringConverterBase__State {

	ret := newQStringConverterBase__State(QStringConverterBase__State_new2(f))
	ret.isSubclass = true
	return ret
}

func (this *QStringConverterBase__State) Clear() {
	QStringConverterBase__State_Clear(this.h)
}

func (this *QStringConverterBase__State) Reset() {
	QStringConverterBase__State_Reset(this.h)
}
